File: game_plugins/hase_und_igel_new/shared/sc/plugin2018/Action.java
package sc.plugin2018;

import com.thoughtworks.xstream.annotations.XStreamAlias;
import com.thoughtworks.xstream.annotations.XStreamAsAttribute;
import sc.shared.InvalidMoveException;

@XStreamAlias(value = "action")
public abstract class Action implements Comparable<Action> {

  public Action() {
    this.order = 0;
  }

  /**
   * Nummer der Aktion. Aktionen werden aufsteigend sortiert nach ihrer Nummer
   * ausgeführt.

   */
  @XStreamAsAttribute
  public int order;

  public abstract void perform(GameState state, Player player) throws InvalidMoveException;

  @Override
  public int compareTo(Action o) {
    return Integer.compare(this.order, o.order);
  }
}

File: plugin/src/main/kotlin/sc/plugin2024/Action.kt
package sc.plugin2024

import com.thoughtworks.xstream.annotations.XStreamAlias
import sc.shared.IMoveMistake

@XStreamAlias(value = "action")
interface Action {
    fun perform(state: GameState): IMoveMistake?
}

interface Addable<T> {
    operator fun plus(other: T): T
}
